modules/backendscheduler/backendscheduler.go

```diff
@@ -0,0 +1,391 @@
+func (s *BackendScheduler) ScheduleOnce(ctx context.Context) error {
+	s.work.Prune()
+
+	for _, job := range s.compactions(ctx) {
+		if err := s.createCompactionJob(ctx, job.Tenant, job.Compaction.Input); err != nil {
+			return fmt.Errorf("failed to create compaction job: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// CreateJob creates a new job
+func (s *BackendScheduler) CreateJob(_ context.Context, j *work.Job) error {
+	err := s.work.AddJob(j)
+	if err != nil {
+		return fmt.Errorf("failed to create job: %w", err)
+	}
+
+	activeCount := 0
+	for _, jj := range s.work.Jobs() {
+		if !jj.IsComplete() && !jj.IsFailed() {
+			activeCount++
+		}
+	}
+
+	metricJobsCreated.WithLabelValues(j.JobDetail.Tenant, j.Type.String()).Inc()
+	metricJobsActive.WithLabelValues(j.JobDetail.Tenant, j.Type.String()).Inc()
+
+	level.Info(log.Logger).Log("msg", "created new job", "job_id", j.ID, "tenant", j.JobDetail.Tenant, "type", j.Type.String())
+	return nil
+}
+
+// GetJob returns a job by ID
+func (s *BackendScheduler) GetJob(_ context.Context, id string) (*work.Job, error) {
+	j := s.work.GetJob(id)
+	if j == nil {
+		return nil, fmt.Errorf("job not found: %s", id)
+	}
+
+	return j, nil
+}
+
+// CompleteJob marks a job as completed
+func (s *BackendScheduler) CompleteJob(_ context.Context, id string) error {
+	j := s.work.GetJob(id)
+	if j == nil {
+		return fmt.Errorf("job not found: %s", id)
+	}
+
+	j.Complete()
+	metricJobsCompleted.WithLabelValues(j.JobDetail.Tenant, j.Type.String()).Inc()
+	metricJobsActive.WithLabelValues(j.JobDetail.Tenant, j.Type.String()).Dec()
+
+	level.Info(log.Logger).Log("msg", "completed job", "job_id", id, "tenant", j.JobDetail.Tenant, "type", j.Type.String())
+	return nil
+}
+
+// FailJob marks a job as failed
+func (s *BackendScheduler) FailJob(_ context.Context, id string) error {
+	j := s.work.GetJob(id)
+	if j == nil {
+		return fmt.Errorf("job not found: %s", id)
+	}
+
+	j.Fail()
+	metricJobsFailed.WithLabelValues(j.JobDetail.Tenant, j.Type.String()).Inc()
+	metricJobsActive.WithLabelValues(j.JobDetail.Tenant, j.Type.String()).Dec()
+
+	level.Info(log.Logger).Log("msg", "failed job", "job_id", id, "tenant", j.JobDetail.Tenant, "type", j.Type.String())
+	return nil
+}
+
+// Next implements the BackendSchedulerServer interface
+func (s *BackendScheduler) Next(_ context.Context, req *tempopb.NextJobRequest) (*tempopb.NextJobResponse, error) {
+	// Find jobs that already exist for this worker
+	for id, j := range s.work.Jobs() {
+		if j.Status() == work.JobStatusPending || j.Status() == work.JobStatusRunning {
+			if j.WorkerID() == req.WorkerId {
+				resp := &tempopb.NextJobResponse{
+					JobId:  j.ID,
+					Type:   j.Type,
+					Detail: j.JobDetail,
+				}
+
+				level.Info(log.Logger).Log("msg", "assigned previous job to worker", "job_id", id, "worker", req.WorkerId)
+				return resp, nil
+			}
+		}
+	}
+
+	// Find next available job
+	for id, j := range s.work.Jobs() {
```

Comment:
similar thought here. this func would be super clear if it read:
```
func() {

j := jobForWorker()

if j == nil {
  j = jobForType()
}

...

}
```

Reply:
This has been updated and reads a little clearer now.